### PR TITLE
Start testing Sociomantic ocean in pipeline

### PR DESCRIPTION
**Pull request by mihails-strasuns**
Recently (https://github.com/sociomantic-tsunami/ocean/pull/321) we have enabled CI testing with vanilla upstream DMD compiler for WIP ocean major branch, as 2.077 provided necessary `Throwable.message()`. 

Would appreciate getting it to the list of automatically tested projects ;)

### CODE REVIEW

**Comment by mihails-strasuns** [comment]
`./src/ocean/text/util/StringSearch.d(265): Deprecation: cannot subtract pointers to different types: `void*` and `const(char)*` used to work with 2.077 :(

**Comment by mihails-strasuns** [comment]
Good, now fails because ocean is built by default with deprecations-as-errors, which, of course, is not suitable for regression testing with upstream. Will require small tweak in our makefile.

**Comment by MartinNowak** [comment]
Nice, will have a look at pre-installing the dependencies with ansible, assuming that they don't change too often.

**Comment by mihails-strasuns** [comment]
> Nice, will have a look at pre-installing the dependencies with ansible, assuming that they don't change too often.

Yes, I think at this point they change less often than once a year. Alternatively, you can enable docker on the CI server and script can fetch the very same image we use for our own CI tests.

**Comment by mihails-strasuns** [comment]
No problem, whatever suits you best. Btw there is no need to add workarounds for deprecation bit - I will update the PR as soon as https://github.com/sociomantic-tsunami/ocean/pull/374 is merged which should be pretty soon :)

File: vars/runPipeline.groovy [hunk 1]
```diff
@@ -206,6 +209,15 @@ def testDownstreamProject (name) {
                     sh 'dub test --compiler=$DC'
                     break;
 
+                case 'sociomantic-tsunami/ocean':
+                    sh '''
+                    sed -i '/^override DFLAGS += -de$/d' Build.mak
+                    git submodule update --init
+                    make d2conv V=1
+                    make test V=1 DVER=2 F=production LDFLAGS=-L./libebtree/usr/lib/
```
Comment (mihails-strasuns) [inline vars/runPipeline.groovy]: Shouldn't need `LDFLAGS` here if dependencies are pre-installed

**Comment by mihails-strasuns** [comment]
Tagged https://github.com/sociomantic-tsunami/ocean/releases/tag/v4.0.0-alpha.1 and pushed commit to switch to it

**Comment by MartinNowak** [comment]
Thx, I'd prefer help with the failing tests.
https://ci.dlang.io/blue/organizations/jenkins/dlang-org%2Fci/detail/PR-88/8/pipeline/188#step-342-log-368
```
ocean.core.Test.TestException@/var/lib/jenkins/dlang_projects@2/sociomantic-tsunami/ocean/test/sysstats/main.d(68): expression 'nan >= 0' evaluates to false
----------------
./src/ocean/core/Enforce.d:416 void ocean.core.Enforce.enforceImpl!(">=", float, int).enforceImpl(lazy Exception, float, int, immutable(char)[], int) [0x6281dd]
./src/ocean/core/Enforce.d:323 void ocean.core.Enforce.enforceImpl!(">=", ocean.core.Test.TestException, float, int).enforceImpl(float, int, immutable(char)[], int) [0x628138]
/var/lib/jenkins/dlang_projects@2/sociomantic-tsunami/ocean/test/sysstats/main.d:68 int 
```

**Comment by mihails-strasuns** [comment]
Can you run the same pipeline with released 2.077 to see if it is because of compiler or environment?

**Comment by mihails-strasuns** [comment]
Though environment is more likely - https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/test/sysstats/main.d#L68

I guess you have virtualized system where CPU number is not accessible?

**Comment by mihails-strasuns** [comment]
Ping @Burgos - I think you are more familiar with stats module than me.

**Comment by MartinNowak** [comment]
Slaves are using LXD or KVM.

**Comment by Burgos** [comment]
I wonder if the problem is simply because the timing scheduling and CPU counting is different in LXD/KVM than on the bare metal machine. We'll try adding some busy loop there to spend some time in the CPU instead of yielding the control back.

**Comment by MartinNowak** [comment]
`Stats(nan, nan, nan, 24403968, 8146944, 0.024311)`
Reproducibly the uptime diff is 0, while there are as many ticks per second (100), lxcfs rounds them to full seconds.
https://insights.ubuntu.com/2015/03/02/introducing-lxcfs/

**Comment by Burgos** [comment]
Hm, ok. Then we could just sleep for two seconds before getting the results. I don't think this will hurt developers much (and they could always `make -j` to speed the tests). Alternatively, we could just ignore this test on the platforms that behave like this.

**Comment by MartinNowak** [comment]
@Burgos see https://github.com/sociomantic-tsunami/ocean/pull/385.

**Comment by mihails-strasuns** [comment]
Updated to 4.0.0-alpha.2, looking at build log

**Comment by mihails-strasuns** [comment]
Green! D-YAML has failed though :)

**Comment by Burgos** [comment]
Such a time to be alive!

**Comment by MartinNowak** [comment]
> Green! D-YAML has failed though :)

Yes, see https://github.com/dlang/dmd/pull/7315#issuecomment-347937938 :/.

**Comment by MartinNowak** [comment]
Let's hope it doesn't run into any other issues on the KVM host, but since it works on Travis-CI/docker, I wouldn't really expect some.

**Comment by mihails-strasuns** [comment]
@MartinNowak I see. Do you want me to have a look into Jenkins instance taking over what Travis/CircleCI currently do?

**Comment by MartinNowak** [comment]
Let's see whether https://github.com/dlang/ci/pull/90 works.

**Comment by Burgos** [comment]
Am I seeing correctly that merging this PR started failing the master? https://ci.dlang.io/blue/organizations/jenkins/dlang-org%2Fci/detail/master/47/pipeline

**Comment by MartinNowak** [comment]
Yes @Burgos, looks like it :/.
Maybe there is sth. wrong with the timer wait loop?
https://ci.dlang.io/blue/organizations/jenkins/dlang-org%2Fci/detail/master/47/pipeline/180#step-313-log-369

Currently rerunning https://ci.dlang.io/blue/organizations/jenkins/dlang-org%2Fci/detail/master/48/pipeline.

**Comment by Burgos** [comment]
The timer loop looks correct to me. I think the problem is that we're checking for equality of the ProcUptime instances: https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/sys/stats/linux/ProcVFS.d#L209 which includes both `uptime` and `idle` but the CPU user calculation uses only a uptime member. https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/sys/Stats.d#L171

**Comment by Burgos** [comment]
(So the `idle` progresses, but not the `uptime`).

**Comment by mihails-strasuns** [comment]
I think for now best course of action would be to disable this one specific test:

```diff
diff --git a/Build.mak b/Build.mak
index 11eb0f0..f875f76 100644
--- a/Build.mak
+++ b/Build.mak
@@ -20,6 +20,10 @@ clean += .*.lst
 TEST_FILTER_OUT += \
        $C/test/collectd/main.d
 
+# fails with KVM, temporarily disabled
+TEST_FILTER_OUT += \
+       $C/test/sysstats/main.d
+
 $O/test-%: override LDFLAGS += -lebtree
 
 $O/test-filesystemevent: override LDFLAGS += -lrt
```

It will fix pipeline for now and we will talk about best way to fix actual test later this week among us.

**Comment by Burgos** [comment]
https://github.com/sociomantic-tsunami/ocean/pull/386

**Comment by mihails-strasuns** [comment]
https://github.com/dlang/ci/pull/91

**Pull request merged**